rma_sale/tests/test_rma_sale.py

```diff
@@ -0,0 +1,229 @@
+
+    def test_create_rma_with_so(self):
+        rma_vals = {
+            "partner_id": self.partner.id,
+            "order_id": self.sale_order.id,
+            "product_id": self.product_1.id,
+            "product_uom_qty": 5,
+            "location_id": self.sale_order.warehouse_id.rma_loc_id.id,
+            "operation_id": self.operation.id,
+        }
+        rma = self.env["rma"].create(rma_vals)
+        rma.action_confirm()
+        self.assertTrue(rma.reception_move_id)
+        self.assertFalse(rma.reception_move_id.origin_returned_move_id)
+
+    @mute_logger("odoo.models.unlink")
+    def test_create_rma_from_so(self):
+        order = self.sale_order
+        wizard = self._rma_sale_wizard(order)
+        rma = self.env["rma"].browse(wizard.create_and_open_rma()["res_id"])
+        self.assertEqual(rma.partner_id, order.partner_id)
+        self.assertEqual(rma.order_id, order)
+        self.assertEqual(rma.picking_id, self.order_out_picking)
+        self.assertEqual(rma.move_id, self.order_out_picking.move_ids)
+        self.assertEqual(rma.product_id, self.product_1)
+        self.assertEqual(rma.product_uom_qty, self.order_line.product_uom_qty)
+        self.assertEqual(rma.product_uom, self.order_line.product_uom)
+        self.assertEqual(rma.state, "confirmed")
+        self.assertEqual(
+            rma.reception_move_id.origin_returned_move_id,
+            self.order_out_picking.move_ids,
+        )
+        self.assertEqual(
+            rma.reception_move_id.picking_id + self.order_out_picking,
+            order.picking_ids,
+        )
+        user = new_test_user(self.env, login="test_refund_with_so")
+        order.user_id = user.id
+        # Receive the RMA
+        rma.action_confirm()
+        rma.reception_move_id.quantity = rma.product_uom_qty
+        rma.reception_move_id.picking_id.button_validate()
+        # Refund the RMA
+        rma.action_refund()
+        self.assertEqual(self.order_line.qty_delivered, 0)
+        self.assertEqual(self.order_line.qty_invoiced, -5)
+        self.assertEqual(rma.refund_id.user_id, user)
+        self.assertEqual(rma.refund_id.invoice_line_ids.sale_line_ids, self.order_line)
+        # Cancel the refund
+        rma.refund_id.button_cancel()
+        self.assertEqual(self.order_line.qty_delivered, 5)
+        self.assertEqual(self.order_line.qty_invoiced, 0)
+        # And put it to draft again
+        rma.refund_id.button_draft()
+        self.assertEqual(self.order_line.qty_delivered, 0)
+        self.assertEqual(self.order_line.qty_invoiced, -5)
+
+    @users("partner@rma")
+    def test_create_rma_from_so_portal_user(self):
+        order = self.sale_order
+        wizard_obj = (
+            self.env["sale.order.rma.wizard"].sudo().with_context(active_id=order.id)
+        )
+        operation = self.rma_operation_model.sudo().search([], limit=1)
+        line_vals = [
+            (
+                0,
+                0,
+                {
```

Comment:
Could you use the [Command](https://github.com/odoo/odoo/blob/7dcdb774e61e6d520d49dc402c4fad6dcf527ac1/odoo/fields.py#L4227-L4341
) class? It’s more readable.

```suggestion
           Command.create(
                {
```